src/main/java/com/medallia/word2vec/neuralnetwork/NeuralNetworkTrainer.java

```diff
@@ -151,28 +156,30 @@ static long incrementRandom(long r) {
 	}
 	
 	/** @return Trained NN model */
-	public NeuralNetworkModel train(Iterable<List<String>> sentences) throws InterruptedException {
-		ListeningExecutorService ex = MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(config.numThreads));
-		
+	public NeuralNetworkModel train(final Iterable<List<String>> sentences) throws InterruptedException {
+		final ListeningExecutorService ex =
+				MoreExecutors.listeningDecorator(
+					new ThreadPoolExecutor(config.numThreads, config.numThreads,
```

Comment:
Neat trick, but let's leave a comment to what we're trying to accomplish here.  If I understand correctly, the overall idea is to have executor.submit block if there are no available threads to avoid materializing the sentences in memory before they are needed.  The ArrayBlockingQueue and CallerRunsPolicy is one way to accomplish this.